app/models/highlight.rb

```diff
@@ -16,54 +21,113 @@ class Highlight < ApplicationRecord
+  def normalize_ids
     if openstax_page?
       self.source_id = source_id&.downcase&.strip
-      source_parent_ids.map!(&:downcase).map!(&:strip)
+      self.scope_id = scope_id&.downcase&.strip
     end
   end
 
-  private
+  protected
+
+  def all_from_scope_and_source
+    Highlight.where(scope_id: scope_id).where(source_id: source_id)
+  end
 
   def valid_uuid?(uuid)
     VALID_UUID.match?(uuid.to_s.downcase)
   end
 
-  def source_ids_uuid_for_openstax_page
-    return unless openstax_page?
+  def validate_uuid(field, is_source_type_specific:)
+    unless VALID_UUID.match?(send(field).to_s.downcase)
+      errors.add(field, "must be a UUID#{' for source_type ' + source_type if is_source_type_specific}")
+    end
+  end
+
+  def must_have_source_type_specific_ids
+    if openstax_page?
+      validate_uuid(:scope_id, is_source_type_specific: true)
+      validate_uuid(:source_id, is_source_type_specific: true)
+    end
+  end
+
+  def must_have_consistent_neighbors
+    if prev_highlight_id.nil? && next_highlight_id.nil? && all_from_scope_and_source.any?
+      errors.add(:base, 'Must specify previous or next highlight because there are other highlights ' \
+                        'in this scope and source.')
+    end
 
-    source_parent_ids.each do |source_parent_id|
-      unless valid_uuid?(source_parent_id)
-        errors.add(:source_parent_ids, 'source parent id of type openstax page must be a uuid')
+    if prev_highlight_id
+      if prev_highlight.nil?
+        errors.add(:prev_highlight_id, 'is unknown')
+      else
+        if prev_highlight.next_highlight_id != next_highlight_id
+          errors.add(:base, 'The specified previous and next highlights are not adjacent')
+        end
+        if prev_highlight.source_id != source_id
+          errors.add(:prev_highlight_id, 'does not reference a highlight in the same source')
+        end
+        if prev_highlight.scope_id != scope_id
+          errors.add(:prev_highlight_id, 'does not live in the same scope')
+        end
       end
     end
-    unless valid_uuid?(source_id)
-      errors.add(:source_id, 'source id of type openstax page must be a uuid')
+
+    if next_highlight_id
+      if next_highlight.nil?
+        errors.add(:next_highlight_id, 'is unknown')
+      else
+        if next_highlight.prev_highlight_id != prev_highlight_id
+          errors.add(:base, 'The specified previous and next highlights are not adjacent')
+        end
+        if next_highlight.source_id != source_id
+          errors.add(:next_highlight_id, 'does not reference a highlight in the same source')
+        end
+        if next_highlight.scope_id != scope_id
+          errors.add(:next_highlight_id, 'does not live in the same scope')
+        end
+      end
     end
   end
 
-  def self.postgres_style_array(souce_parent_ids)
-    "{#{souce_parent_ids.join(',')}}"
+  def insert_new_highlight_between_neighbors
+    self.order_in_source =
+      if prev_highlight && next_highlight
```

Comment:
would be nice if this logic were in a method with a good name that describes what setting this float means.  Perhaps  `def find_relative_order_by_source` and with a short blurb on what would happen if the order_in_source reaches the mathematical end of the float range (if that's realistic). 

(This will definitely be one of those things that will really mess things up if anyone deletes records from postgres and not thru the model)

Reply:
Moved it